### PR TITLE
Migrate entire SDK from RUN to Inter (IST)

### LINKED ISSUE #4800: Remove “BLD” and “RUN” keys from run-protocol code
# RUN/IST role cases

## vaultFactory
Token that is minted, secured by collateral (per vault manager)
keyword: `Minted`
Proposal shape will be {give: Collateral, want: Debt}

## runStake / bldBoost
Token that is minted, secured by collateral.
keyword: 'Minted'

## feeDistributor
Fee.
keyword: `Fee`
The CollectFees offer result will by a seat with `getPayout('Fee')`.

## assetReserve
Fee mint to hold penalties
keyword: `Fee`

## psm
Stable
keyword: Stable (already)

### PR DESCRIPTION
**Pull request by turadg**
closes: #4800
part of: #5326

## Description

Per https://github.com/Agoric/agoric-sdk/issues/4800:
- make vaultFactory and runStake use `Minted` instead of `RUN`
- make feeDistributor and assetReserve use `Fee` instead of `RUN`

This PR has been repurposed to do a "big bang" migration from RUN to Inter.  Here are the gates we need to pass before that can land:

- [x] This base PR is approved (@Chris-Hibbert)
- [x] https://github.com/Agoric/agoric-sdk/pull/5802
- [x] Bootstrap changes from RUN to IST https://github.com/Agoric/agoric-sdk/pull/5332
  - [x] Cosmos-level denom changes from `urun` to `uist`
- [x] Load generator changes to accommodate new keywords/tokens https://github.com/Agoric/testnet-load-generator/pull/80/ (@mhofman)
- [x] make getting-started integration pass against this branch (if necessary)

After this PR:
- [ ] Any necessary manual dapp testing in #5591 has been performed (@samsiegart)
- [ ] Documentation has been updated (@dckc)
- [ ] [run-protocol package](https://www.npmjs.com/package/@agoric/run-protocol) is deprecated and points to inter
- [ ] no results for `git grep "RUN [pP]rotocol"` 

The intention is for green CI to indicate that this job was done correctly.

## References in other repositories

If you are mentioned in a given line below, see <a href="#user-content-updating-references">instructions for updating references</a>.

DO NOT remove your mention in case problems are found that need your attention.  If you have done this correctly, somebody will merge your PR when the big bang happens.

#dapp-fungible-faucet-branch main @turadg 
#dapp-card-store-branch main @turadg
#dapp-oracle-branch: main @turadg
#dapp-otc-branch: main @turadg 
#dapp-pegasus-branch: main @turadg 
#dapp-simple-exchange-branch: main @turadg 
#dapp-treasury-branch: main @turadg 
#documentation-branch: main @turadg
#loadgen-branch: main Agoric/testnet-load-generator#80 @mhofman

### <a id="updating-references">Updating References</a>

You can edit the above lines as follows:

- if you have verified that the appropriate repo does not require any changes
  - Set `TODO-BRANCH` to either `master` or `main` as applicable
  - Set `PRNUM` to `n/a`
- if you need to create a new PR in that repo to enact changes
  - Ensure your PR contains the following in its description and wait until it passes its own CI:
     `agoric-sdk-branch: 4800-unRUN`
  - Set `TODO-BRANCH` to the name of the branch
  - Set `PRNUM` to reference the appropriate PR

### Testing Considerations

Once this PR is sufficiently ready (for example, incorporates #5332), a full end to end integration test can run:
- [x] Add the `force-integration` label
- [x] Verify that the "deployment test" passes
  - Optionally double check the logs that IST was used for Vault and AMM tasks.
- [x] Merge the [loadgen PR](https://github.com/Agoric/testnet-load-generator/pull/80) (it's backwards compatible, and must pass tests on master):
  - [x] Mark the loadgen PR ready for review
  - [x] Re-run the integration tests on that PR to make sure it still works against the SDK's master branch
  - [x] Approve and merge
- [x] Edit the description of this PR to comment out the `#loadgen-branch` instruction
- [ ] Land this

### CODE REVIEW

**Comment by dckc** [comment]
> advances: #4800 (not sure what else may need to be done after this)

This looks on target; let's re-scope 4800 to match this and change "advances" to "closes".

In particular, I can't think of any any `BLD` keywords that are worth changing.

As to

> - Make them code parameters

That's not feasible if it means runtime-changeable parameters. If it means manifest constants, then maybe, but even that is awkward sometimes.

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/reserve/assetReserve.js]: >  * This contract has the ability to tokens on the Fee mint provided in its

 This contract has the ability to mint tokens of the Fee brand provided in its

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/reserve/assetReserve.js]: s/fee/Fee/

Reword so there aren't two uses of "it" that refer to different things.  (Yeah, I know. I probably wrote it originally.)

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/reserve/assetReserve.js]: Please bring back the manifest constant rather than using 'Fee'.

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/vaultFactory/test-liquidator.js]: `run.make`?

File: packages/run-protocol/test/reserve/test-reserve.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/reserve/test-reserve.js]: `oneKRun`?

File: packages/run-protocol/test/reserve/test-reserve.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/reserve/test-reserve.js]: `runBrand`?

**Comment by Chris-Hibbert** [comment]
Okay, I was going to figure out how to review it one commit at a time, but you've re-pushed with different commits when I was half-way through. That makes it more difficult.

File: packages/run-protocol/src/contractSupport.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/contractSupport.js]: This is now a general-purpose helper. Drop the "usually ..." part and add a top comment saying something like "verifies that every key in the proposal is in the provided list".

**Comment by turadg** [comment]
> That makes it more difficult.

Sorry about that. By putting it in Draft I meant to convey that it’s not stable. I just did a comb through and rebased on master  so now it's [ready for review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review).

Until it's approved I'll push fixup commits instead of rebasing. (E.g. one to roll back the Zoe error message change that's failing a test)

File: packages/zoe/test/zoeTestHelpers.js
Comment (Chris-Hibbert) [inline packages/zoe/test/zoeTestHelpers.js]: I'm surprised lint didn't complain. `X` isn't used in this file.  (at least in  this commit)

File: packages/run-protocol/src/vaultFactory/AttackersGuide.md
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/AttackersGuide.md]: Why curly braces? gitHub's preview doesn't render that specially. Are you expecting some other tool to add emphasis?  If the contract uses `Minted`, just use that, I think.

File: packages/run-protocol/src/vpool-xyk-amm/constantProduct/swap.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/constantProduct/swap.js]: I don't find the curly braces helpful. If you want emphasis, I think I'd use `*Fee*`, but I think the capital letter is sufficient when used consistently.

File: packages/run-protocol/src/vpool-xyk-amm/doublePool.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vpool-xyk-amm/doublePool.js]: curly braces.

**Review by Chris-Hibbert — COMMENTED** [review]
Here are some reviews. I'll get them in before anything more changes.  Not done yet.

**Review by michaelfig — DISMISSED** [review]
As discussed in standup today, I'm blocking the merge of this PR until the rest of the platform RUN->IST changes can be merged into this one for a "big-bang" release.

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (turadg) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: Yes, the fee is in RUN.

File: packages/run-protocol/test/reserve/test-reserve.js
Comment (turadg) [reply packages/run-protocol/test/reserve/test-reserve.js]: Same, as above, the fee in the contract instance is in RUN. Do you think it should be IST instead?

IMO, IST vs RUN is a bootstrap concern and not of the protocol itself. It should be capable of being wired up with any brand and having some tests use RUN helps ensure it.

File: packages/zoe/test/zoeTestHelpers.js
Comment (turadg) [reply packages/zoe/test/zoeTestHelpers.js]: lint job did complain

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (turadg) [reply packages/run-protocol/src/reserve/assetReserve.js]: Porque? It's not something that needs to be refactored. It's less Jessie-safe and @dckc has back-tracked on using them for keywords.

File: packages/run-protocol/src/vaultFactory/AttackersGuide.md
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/AttackersGuide.md]: to convey that it's a named parameter of the contract. if you think "borrow Minted in exchange" is preferred I'm open to that, but want to hold off until there's some more discussion. e.g. whether to backtick quote them. It's easy to globally change `{Minted}` but harder for `Minted`.

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (dckc) [reply packages/run-protocol/src/reserve/assetReserve.js]: hm. we weren't 100% thorough about `RunKW`; for example: `const offerToSeat = runMint.mintGains(harden({ RUN: runAmount }))`.

I can see how `FeeKW` would be better than a string constant in cases such as `saveBrandKeyword(feeKit.brand, 'Fee');` but do we like `proposal.give[FeeKW]` better than `proposal.give.Fee`? When I tried it, I didn't find that the result was worth a wholesale style change to the `proposal.give[FeeKW]` idiom.

Comment (michaelfig) [reply packages/run-protocol/src/reserve/assetReserve.js]: BTW, @turadg, using variable keywords for object literals should be supported by Jessie.  If it isn't, that's a bug we need to fix in the Jessie spec, not something you need to avoid here.

**Comment by michaelfig** [comment]
Tagging @Agoric/internal for visibility; unsubscribe if you're not interested.

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (turadg) [reply packages/run-protocol/src/reserve/assetReserve.js]: > using variable keywords for object literals should be supported by Jessie

By "Jessie-safe" I meant passes `@jessie-check`, which causes this lint error :
Screen Shot 2022-06-14 at 11 53 55 AM
 
Are you saying that's a bug in the Jessie linter?

To be clear, while that's a discussion of the drawback of using a manifest constant for keyword my main reason for pushing back on doing the keywords is it's more work (delaying release) for at best negligible value.

**Comment by michaelfig** [comment]
@Agoric/internal see [the description above](https://github.com/Agoric/agoric-sdk/pull/5585#issue-1270073769).

File: packages/run-protocol/src/vaultFactory/liquidation.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/liquidation.js]: s/runToBurn/mintedToBurn/

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/vault.js]: s/assertVaultHoldsNoRun/assertVaultHoldsNoMinted/

What happened to the review button for "show how to change the code"?

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/vault.js]: s/wantRUN/wantMinted/

File: packages/run-protocol/src/vaultFactory/vaultFactory.js
Comment (Chris-Hibbert) [inline packages/run-protocol/src/vaultFactory/vaultFactory.js]: s/runIssuerRecord:/wantedIssuerRecord:/

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (Chris-Hibbert) [inline packages/run-protocol/test/vaultFactory/test-liquidator.js]: s/expectedRUN/expectedWanted/

File: packages/run-protocol/src/reserve/assetReserve.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/reserve/assetReserve.js]: > Please bring back the manifest constant rather than using 'Fee'.

I take it back.  We use the replacements for RUN as strings and keywords directly throughout much of this code, and it would be inconsistent to only require the manifest constant in Reserve, and a hassle to switch all of InterProtocol to manifest constants.

File: packages/run-protocol/src/vaultFactory/AttackersGuide.md
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vaultFactory/AttackersGuide.md]: The curly braces don't do it for me. I wouldn't mind backticks.

File: packages/run-protocol/test/reserve/test-reserve.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/reserve/test-reserve.js]: I didn't get the "as above" reference.  

Are you saying that "Run" in tests is intended as an arbitrary currency, no different than "moola" or "simoleans"?  I don't find it as evocative of money as our other common terms.

File: packages/run-protocol/src/vaultFactory/liquidation.js
Comment (dckc) [reply packages/run-protocol/src/vaultFactory/liquidation.js]: Reasonable suggestion, but also above and beyond the call of duty. `runToBurn` isn't visible via the contract API.

File: packages/run-protocol/src/vaultFactory/vault.js
Comment (turadg) [reply packages/run-protocol/src/vaultFactory/vault.js]: > What happened to the review button for "show how to change the code"?

It's only available when reviewing the PR changes combined. When viewing per commit they can't guarantee what the _before_ is that you're proposing to edit.

File: packages/run-protocol/test/reserve/test-reserve.js
Comment (turadg) [reply packages/run-protocol/test/reserve/test-reserve.js]: > I didn't get the "as above" reference.

https://github.com/Agoric/agoric-sdk/pull/5585#discussion_r897177412

> Are you saying that "Run" in tests is intended as an arbitrary currency, no different than "moola" or "simoleans"?

Yes.

> I don't find it as evocative of money as our other common terms.

Granted. The other terms would be marginally better. But changing to them (or IST) is additional work that I don't think is warranted.

File: packages/run-protocol/test/vaultFactory/test-liquidator.js
Comment (turadg) [reply packages/run-protocol/test/vaultFactory/test-liquidator.js]: the test uses RUN brand so I think this is still valid, just as `expectedAEth` is

**Review by dckc — COMMENTED** [review]
I gather this is not currently ready for review. taking it off my queue for now...

File: packages/agoric-cli/src/chain-config.js
Comment (Chris-Hibbert) [inline packages/agoric-cli/src/chain-config.js]: This reads as if it's equal to USD. I'd guess we want some broad disclaimers if we're even mentioning USD. I might say that it "generally tracks USD".

File: packages/inter-protocol/src/contractSupport.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/contractSupport.js]: I would drop the "usually" clause. It most be left over from before this was a generic helper.

File: packages/inter-protocol/src/feeDistributor.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/feeDistributor.js]: that's a random drive-by correction. What justifies including it with this PR? Does it deserve a test?

File: packages/inter-protocol/src/proposals/demoIssuers.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/proposals/demoIssuers.js:306]: The old format makes it easier to see that the two tests are parallel.    Just a suggestion

File: packages/inter-protocol/src/reserve/assetReserve.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/reserve/assetReserve.js]:
```suggestion
      `'Fee' is a special case handled by the reserve contract`,
```

File: packages/inter-protocol/src/runStake/runStakeKit.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/runStake/runStakeKit.js]: "holds no debt" sounds colloquial. how about `holdsNoDebtTokens`?

File: packages/inter-protocol/src/runStake/runStakeKit.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/runStake/runStakeKit.js:324]: Everywhere else in this file it's **Debt**. Why does this case turn into **Minted**?

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/vaultFactory/vault.js]:
```suggestion
  getMintedAllocated: ({ facets }, seat) =>
```

File: packages/inter-protocol/src/vaultFactory/vault.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/vaultFactory/vault.js]:
```suggestion
  assertVaultHoldsNoMinted: ({ state, facets }) => {
```

File: packages/inter-protocol/src/vpool-xyk-amm/doublePool.js
Comment (Chris-Hibbert) [inline packages/inter-protocol/src/vpool-xyk-amm/doublePool.js]: This feels like implementation doc rather than API description.  I think it makes more sense to move it outside the jsdoc. If you want some description for the jsdoc, I'd say

> doublePool is the virtualPool implementation for calculating prices when transiting through two pools. virtual pools wrap something that can do a swap and calculate prices.  Current wrapped pools are single pools and pairs of pools, but we've also contemplated using a similar virtual wrapper for other curves.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/inter-protocol/src/feeDistributor.js
Comment (turadg) [reply packages/inter-protocol/src/feeDistributor.js]: > What justifies including it with this PR? 

IIRC it's because I ran into an error that this would have helped me diagnose. So error detection.

> Does it deserve a test?

No because it's used only for `distributeFees()` which requires a real Payment. It's also ostensibly what `collectFees` returns.

But I'm not wed to it. I'll remove to keep this PR cleaner.

File: packages/inter-protocol/src/proposals/demoIssuers.js
Comment (turadg) [reply packages/inter-protocol/src/proposals/demoIssuers.js:306]: Agreed. The problem is that with `[Stake.symbol, Stable.symbol].includes(issuerName)` TypeScript complains,
```
Argument of type 'string' is not assignable to parameter of type '"BLD" | "IST"'.
```
This is because `Stake.symbol` is `'BLD'` instead of `string` that `'BLD'` was.

File: packages/inter-protocol/src/runStake/runStakeKit.js
Comment (turadg) [reply packages/inter-protocol/src/runStake/runStakeKit.js]: done

File: packages/inter-protocol/src/runStake/runStakeKit.js
Comment (turadg) [reply packages/inter-protocol/src/runStake/runStakeKit.js:324]: IIRC just an incomplete move from Debt terminology.

File: packages/inter-protocol/src/vpool-xyk-amm/doublePool.js
Comment (turadg) [reply packages/inter-protocol/src/vpool-xyk-amm/doublePool.js]: done

**Review by Chris-Hibbert — APPROVED** [review]

**Pull request merged**
